models/middleware.js

```diff
@@ -0,0 +1,296 @@
+    Middleware.create(phasePlaceHolder, cb);
+  });
+};
+
+Middleware.addMiddleware = function(m, cb) {
+  var self = this;
+  this.addPhase(m.facetName, m.phase, m.nextPhase, function(err, p) {
+    if (err) return cb(err);
+    if (m.isPhasePlaceHolder) return cb(null, p);
+    self.create(m, cb);
+  });
+};
+
+function compareByOrder(m1, m2) {
+  if (m1.order == null && typeof m2.order === 'number') {
+    return 1;
+  }
+  if (m2.order == null && typeof m1.order === 'number') {
+    return -1;
+  }
+  var diff = m1.order - m2.order;
+  if (isNaN(diff) || diff === 0) {
+    diff = 0;
+  }
+  return diff > 0 ? 1 : -1;
+}
+
+var subPhaseOrders = {
+  before: 1,
+  regular: 2,
+  after: 3
+};
+
+function compareBySubPhase(m1, m2) {
+  var sp1 = m1.subPhase || 'regular';
+  var sp2 = m2.subPhase || 'regular';
+  return subPhaseOrders[sp1] - subPhaseOrders[sp2];
+}
+
+function sortMiddleware(instances) {
+  // Find all phases
+  var phases = instances.filter(function(m) {
+    return m.isPhasePlaceHolder;
+  });
+
+  // Find regular middleware entries
+  var middleware = instances.filter(function(m) {
+    return !m.isPhasePlaceHolder;
+  });
+
+  // Sort the entries to keep the order
+  stableSortInPlace(phases, compareByOrder);
+
+  // Build a map for phase orders (phaseName --> phaseOrder)
+  var phaseOrders = {};
+  phases.forEach(function(p) {
+    phaseOrders[p.phase] = p.order;
+  });
+
+  stableSortInPlace(middleware, function(m1, m2) {
+    // First by phase
+    var delta = phaseOrders[m1.phase] - phaseOrders[m2.phase];
+    if (delta !== 0) return (delta > 0 ? 1 : -1);
+    // by subPhase
+    delta = compareBySubPhase(m1, m2);
+    if (delta !== 0) return (delta > 0 ? 1 : -1);
+    // By order
+    return compareByOrder(m1, m2);
+  });
+  return {
+    phases: phases,
+    middleware: middleware
+  };
+}
+
+/**
+ * Load all middleware instances from cache and sort them by order
+ * @param cache
+ */
+function loadFromCache(cache) {
+  var instances = Middleware.allFromCache(cache);
+  var results = sortMiddleware(instances);
+  var phases = results.phases;
+  var middleware = results.middleware;
+  phases.forEach(function(p) {
+    var entries = middleware.filter(function(m) {
+      return p.phase === m.phase;
+    });
+    entries.forEach(function(m) {
+      var subPhase = m.subPhase || 'regular';
+      p[subPhase] = p[subPhase] || [];
+      p[subPhase].push(m);
+    });
+  });
+  return phases;
+}
+
+/**
+ * Serialize the middleware model instances to the JSON object for
+ * middleware.json
+ * @param {*[]} cache The cache data source
+ * @param {String} facetName Facet name
+ * @returns {ConfigFile}
+ */
+Middleware.serialize = function(cache, facetName) {
+  var middlewarePath = path.join(facetName, 'middleware.json');
+  var phases = loadFromCache(cache);
+  var middlewareConfig = {};
+
+  function addEntries(phase, subPhase, entries) {
+    if (Array.isArray(entries)) {
```

Comment:
Nitpick: use inverse condition + early return, it makes the code easier to understand and less nested.

```js
if (!Array.isArray(entries) return 0;
entries.forEach(function(m)  {
  // ...
});
return entries.length;
```